Celeste.Mod.mm/Mod/Entities/CustomTextVignette.cs

```diff
@@ -0,0 +1,188 @@
+                    }
+                }
+                yield return 2.5f;
+
+                while (textAlpha > 0f) {
+                    textAlpha -= 1f * Engine.DeltaTime;
+                    yield return null;
+                }
+
+                textAlpha = 0f;
+                textStart = text.GetNextPageStart(textStart);
+                yield return 0.5f;
+            }
+            textStart = int.MaxValue;
+        }
+
+        public override void Update() {
+            if (menu == null) {
+                base.Update();
+                if (!exiting) {
+                    if (textCoroutine != null && textCoroutine.Active) {
+                        textCoroutine.Update();
+                    }
+
+                    timer += Engine.DeltaTime;
+                    if (timer >= SFXDuration && !started) {
+                        StartGame();
+                    }
+                    if (timer < (SFXDuration - 2f) && menu == null && (Input.Pause.Pressed || Input.ESC.Pressed)) {
+                        OpenMenu();
+                    }
+                }
+            } else if (!exiting) {
+                menu.Update();
+            }
+
+            pauseFade = Calc.Approach(pauseFade, (menu != null) ? 1 : 0, Engine.DeltaTime * 8f);
+            renderer.BackgroundFade = Calc.Approach(renderer.BackgroundFade, (menu != null) ? 0.6f : 0f, Engine.DeltaTime * 3f);
+            fade = Calc.Approach(fade, 0f, Engine.DeltaTime);
+        }
+
+        public void OpenMenu() {
+            Audio.Play(SFX.ui_game_pause);
+            Audio.Pause(sfx);
+            Add(menu = new TextMenu());
+            menu.Add(new TextMenu.Button(Dialog.Clean("intro_vignette_resume")).Pressed(CloseMenu));
+            menu.Add(new TextMenu.Button(Dialog.Clean("intro_vignette_skip")).Pressed(StartGame));
+            menu.Add(new TextMenu.Button(Dialog.Clean("intro_vignette_quit")).Pressed(ReturnToMap));
+            menu.OnCancel = menu.OnESC = menu.OnPause = CloseMenu;
+        }
+
+        private void CloseMenu() {
+            Audio.Play(SFX.ui_game_unpause);
+            Audio.Resume(sfx);
+            menu?.RemoveSelf();
+            menu = null;
+        }
+
+        private void StartGame() {
+            textCoroutine = null;
+            StopSfx();
+            session.Audio.Music.Event = areaMusic;
+
+            if (menu != null) {
+                menu.RemoveSelf();
+                menu = null;
+            }
+
+            new FadeWipe(this, false, () => Engine.Scene = new LevelLoader(session))
+                .OnUpdate = (f) => textAlpha = Math.Min(textAlpha, 1f - f);
+
+            started = true;
+            exiting = true;
+        }
+
+        private void ReturnToMap() {
+            StopSfx();
+            menu.RemoveSelf();
+            menu = null;
+            exiting = true;
+            bool toAreaQuit = ((patch_SaveData) SaveData.Instance).Areas_Safe[0].Modes[0].Completed && Celeste.PlayMode != Celeste.PlayModes.Event;
```

Comment:
It should be possible to clean up the cast + \_Safe access to `SaveData.Instance.Areas`